base/test/launcher/teamcity_service_messages.cc

```diff
@@ -0,0 +1,119 @@
+/* Copyright (c) 2024 The Brave Authors. All rights reserved.
+ * This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this file,
+ * You can obtain one at https://mozilla.org/MPL/2.0/. */
+
+#include "brave/base/test/launcher/teamcity_service_messages.h"
+
+#include <string>
+
+#include "base/ranges/algorithm.h"
+#include "base/strings/string_number_conversions.h"
+
+namespace base {
+namespace {
+
+// https://www.jetbrains.com/help/teamcity/service-messages.html#Escaped+Values
+std::string_view TeamcityValueEscape(std::string_view s, std::string& result) {
+  static constexpr char kSymbolsToEscape[] = "\n\r'|[]";
+  const char* s_char = ranges::find_first_of(s, kSymbolsToEscape);
+  if (s_char == s.end()) {
+    return s;
+  }
+
+  result.reserve(s.length() + s.length() / 4);
+  result.assign(s.begin(), s_char);
+
+  for (; s_char != s.end(); ++s_char) {
+    switch (*s_char) {
+      case '\n':
+        result += "|n";
+        break;
+      case '\r':
+        result += "|r";
+        break;
+      case '\'':
+      case '|':
+      case '[':
+      case ']':
+        result += '|';
+        [[fallthrough]];
+      default:
+        result += *s_char;
+    }
+  }
+
+  return result;
+}
+
+}  // namespace
+
+TeamcityServiceMessages::Message::Message(std::ostream& ostream,
+                                          std::string_view name)
+    : ostream_(ostream) {
+  (*ostream_) << "##teamcity[" << name;
+}
+
+TeamcityServiceMessages::Message::~Message() {
+  (*ostream_) << "]" << std::endl;
+}
+
+TeamcityServiceMessages::Message&
+TeamcityServiceMessages::Message::WriteProperty(std::string_view name,
+                                                std::string_view value) {
+  if (!value.empty()) {
+    std::string escaped_value;
+    (*ostream_) << " " << name << "='"
```

Comment:
Maybe add escaping for name?
P.S. The current code doesn't need escaping of `name`, because it always passes const char[]. It's just for consistency or/and future changes.

Reply:
`name` cannot be set externally, we control this value, so there's no code path where it should be escaped.

I've moved `TeamcityServiceMessages::Message` class into `private` to make it clear that it's not supposed to be used directly.